internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -736,3 +782,72 @@ func getComposerResourcesNames(cds []ComposedResource) []string {
 	}
 	return names
 }
+
+type compositionResultMeta struct {
+	numWarningEvents   int
+	conditionTypesSeen map[xpv1.ConditionType]struct{}
+}
+
+func (r *Reconciler) handleCommonCompositionResult(ctx context.Context, res CompositionResult, xr *composite.Unstructured, log logging.Logger) compositionResultMeta {
+	cm, err := getClaimFromXR(ctx, r.client, xr)
+	if err != nil {
+		log.Debug(errGetClaim, "error", err)
+	}
+
+	numWarningEvents := 0
+	for _, e := range res.Events {
+		if e.Event.Type == event.TypeWarning {
+			numWarningEvents++
+		}
+
+		detailedMessage := e.Detail + e.Message
+		log.Debug(detailedMessage)
+		r.record.Event(xr, event.Event{
+			Type:        e.Type,
+			Reason:      e.Reason,
+			Message:     detailedMessage,
+			Annotations: e.Annotations,
+		})
```

Comment:
Might be a little neater to add methods on `TargetedEvent` to produce the appropriate events. e.g.:

```go
type TargetedEvent struct {
	event.Event
	Target      CompositionTarget
	Detail      string
}

func (e *TargetedEvent) AsEvent() event.Event {
	return event.Event{Type: e.Type, Reason: e.Reason, Message: e.Message, Annotations: e.Annotations}
}

func (e *TargetedEvent) AsDetailedEvent() event.Event {
	msg := fmt.Sprintf("%s: %s", e.Detail, e.Message)
	return event.Event{Type: e.Type, Reason: e.Reason, Message: msg, Annotations: e.Annotations}
}
```

Reply:
This does look nicer. I added this with a slight change to `AsDetailedEvent`.